{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "bedf0bc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "a9a4ab71",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Функция сохранения заголовка формата VEP\n",
    "def make_header(file_name):\n",
    "    header_list = []\n",
    "    with open(file_name, \"r\") as file:\n",
    "        while True:\n",
    "            line = file.readline()\n",
    "            if not line:\n",
    "                break\n",
    "            if line.split()[0] == \"#Uploaded_variation\": # сохраним заголовок\n",
    "                header_list.append(line.split())\n",
    "                return header_list[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "8cfd407f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>Extra</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CM000663.2_46248360_A/C</td>\n",
       "      <td>CM000663.2:46248360</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000085999</td>\n",
       "      <td>ENST00000371975</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>598</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;STRAND=1;VARIANT_CLASS=SNV;SYM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CM000663.2_46248360_A/C</td>\n",
       "      <td>CM000663.2:46248360</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000085999</td>\n",
       "      <td>ENST00000442598</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>69</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;STRAND=1;VARIANT_CLASS=SNV;SYM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CM000663.2_46248360_A/C</td>\n",
       "      <td>CM000663.2:46248360</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000085999</td>\n",
       "      <td>ENST00000463715</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;STRAND=1;FLAGS=cds_end_NF;VARI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CM000663.2_46248360_A/C</td>\n",
       "      <td>CM000663.2:46248360</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000085999</td>\n",
       "      <td>ENST00000469835</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;STRAND=1;FLAGS=cds_end_NF;VARI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CM000663.2_46248360_A/C</td>\n",
       "      <td>CM000663.2:46248360</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000085999</td>\n",
       "      <td>ENST00000487700</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODIFIER;DISTANCE=152;STRAND=1;VARIANT_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #Uploaded_variation             Location Allele             Gene  \\\n",
       "0  CM000663.2_46248360_A/C  CM000663.2:46248360      C  ENSG00000085999   \n",
       "1  CM000663.2_46248360_A/C  CM000663.2:46248360      C  ENSG00000085999   \n",
       "2  CM000663.2_46248360_A/C  CM000663.2:46248360      C  ENSG00000085999   \n",
       "3  CM000663.2_46248360_A/C  CM000663.2:46248360      C  ENSG00000085999   \n",
       "4  CM000663.2_46248360_A/C  CM000663.2:46248360      C  ENSG00000085999   \n",
       "\n",
       "           Feature Feature_type            Consequence cDNA_position  \\\n",
       "0  ENST00000371975   Transcript    5_prime_UTR_variant           598   \n",
       "1  ENST00000442598   Transcript    5_prime_UTR_variant            69   \n",
       "2  ENST00000463715   Transcript         intron_variant             -   \n",
       "3  ENST00000469835   Transcript         intron_variant             -   \n",
       "4  ENST00000487700   Transcript  upstream_gene_variant             -   \n",
       "\n",
       "  CDS_position Protein_position Amino_acids Codons Existing_variation  \\\n",
       "0            -                -           -      -                  -   \n",
       "1            -                -           -      -                  -   \n",
       "2            -                -           -      -                  -   \n",
       "3            -                -           -      -                  -   \n",
       "4            -                -           -      -                  -   \n",
       "\n",
       "                                               Extra  \n",
       "0  IMPACT=MODIFIER;STRAND=1;VARIANT_CLASS=SNV;SYM...  \n",
       "1  IMPACT=MODIFIER;STRAND=1;VARIANT_CLASS=SNV;SYM...  \n",
       "2  IMPACT=MODIFIER;STRAND=1;FLAGS=cds_end_NF;VARI...  \n",
       "3  IMPACT=MODIFIER;STRAND=1;FLAGS=cds_end_NF;VARI...  \n",
       "4  IMPACT=MODIFIER;DISTANCE=152;STRAND=1;VARIANT_...  "
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Чтение VEP файла\n",
    "vep_file_path = 'DS44625.vep.vcf'\n",
    "header_list = make_header(vep_file_path)\n",
    "vep_df = pd.read_table(vep_file_path, header=None, comment='#')\n",
    "vep_df.columns = header_list\n",
    "\n",
    "# Вывод первых нескольких строк данных\n",
    "vep_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "17a5be4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'IMPACT=MODIFIER;STRAND=1;VARIANT_CLASS=SNV;SYMBOL=RAD54L;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:9826;BIOTYPE=protein_coding;CANONICAL=YES;MANE_SELECT=NM_003579.4;TSL=1;APPRIS=P1;CCDS=CCDS532.1;ENSP=ENSP00000361043;SWISSPROT=Q92698.176;UNIPARC=UPI0000378007;GENE_PHENO=1;EXON=1/18;HGVSc=ENST00000371975.9:c.-46A>C'"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "#vep_df['Consequence']\n",
    "vep_df['Extra'][0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "55f4df7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 11592 entries, 0 to 11591\n",
      "Data columns (total 14 columns):\n",
      " #   Column               Non-Null Count  Dtype \n",
      "---  ------               --------------  ----- \n",
      " 0   #Uploaded_variation  11592 non-null  object\n",
      " 1   Location             11592 non-null  object\n",
      " 2   Allele               11592 non-null  object\n",
      " 3   Gene                 11592 non-null  object\n",
      " 4   Feature              11592 non-null  object\n",
      " 5   Feature_type         11592 non-null  object\n",
      " 6   Consequence          11592 non-null  object\n",
      " 7   cDNA_position        11592 non-null  object\n",
      " 8   CDS_position         11592 non-null  object\n",
      " 9   Protein_position     11592 non-null  object\n",
      " 10  Amino_acids          11592 non-null  object\n",
      " 11  Codons               11592 non-null  object\n",
      " 12  Existing_variation   11592 non-null  object\n",
      " 13  Extra                11592 non-null  object\n",
      "dtypes: object(14)\n",
      "memory usage: 1.2+ MB\n"
     ]
    }
   ],
   "source": [
    "vep_df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "f931a095",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Словарь для перевода названий хромосом\n",
    "chromosome_dict_hg38 = {\n",
    "    \"CM000663.2\": \"chr1\",\n",
    "    \"CM000664.2\": \"chr2\",\n",
    "    \"CM000665.2\": \"chr3\",\n",
    "    \"CM000666.2\": \"chr4\",\n",
    "    \"CM000667.2\": \"chr5\",\n",
    "    \"CM000668.2\": \"chr6\",\n",
    "    \"CM000669.2\": \"chr7\",\n",
    "    \"CM000670.2\": \"chr8\",\n",
    "    \"CM000671.2\": \"chr9\",\n",
    "    \"CM000672.2\": \"chr10\",\n",
    "    \"CM000673.2\": \"chr11\",\n",
    "    \"CM000674.2\": \"chr12\",\n",
    "    \"CM000675.2\": \"chr13\",\n",
    "    \"CM000676.2\": \"chr14\",\n",
    "    \"CM000677.2\": \"chr15\",\n",
    "    \"CM000678.2\": \"chr16\",\n",
    "    \"CM000679.2\": \"chr17\",\n",
    "    \"CM000680.2\": \"chr18\",\n",
    "    \"CM000681.2\": \"chr19\",\n",
    "    \"CM000682.2\": \"chr20\",\n",
    "    \"CM000683.2\": \"chr21\",\n",
    "    \"CM000684.2\": \"chr22\",\n",
    "    \"CM000685.2\": \"chrX\",\n",
    "    \"CM000686.2\": \"chrY\"\n",
    "}\n",
    "\n",
    "def translate_chromosome(cm_id):\n",
    "    return chromosome_dict_hg38.get(cm_id, \"Unknown chromosome\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "9d51a02b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Функция для извлечения гена и генетического варианта из строки 'Extra'\n",
    "def extract_gene_variant(extra_info):\n",
    "    extra_dict = dict(item.split('=') for item in extra_info.split(';') if '=' in item)\n",
    "    gene = extra_dict.get('SYMBOL', '')\n",
    "    genetic_variant = extra_dict.get('HGVSc', '')\n",
    "    if genetic_variant:\n",
    "        genetic_variant = genetic_variant.split(':', 1)[-1]\n",
    "    return pd.Series([gene, genetic_variant], index=['ГЕН', 'ГЕНЕТИЧЕСКИЙ ВАРИАНТ'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "4fd28e93",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Итератор, который обрабатывает DataFrame построчно\n",
    "def gene_variant_iterator(df):\n",
    "    for index, row in df.iterrows():\n",
    "        gene, genetic_variant = extract_gene_variant(row['Extra'])\n",
    "        location = row['Location'].split(':', 1)[0]\n",
    "        location = translate_chromosome(location)\n",
    "        feature = row['Feature']\n",
    "        yield {'ГЕН': gene, 'ПОЗИЦИЯ': location, 'ГЕНЕТИЧЕСКИЙ ВАРИАНТ': genetic_variant, 'ТРАНСКРИПТ': feature, }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "6809fef7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ГЕН</th>\n",
       "      <th>ПОЗИЦИЯ</th>\n",
       "      <th>ГЕНЕТИЧЕСКИЙ ВАРИАНТ</th>\n",
       "      <th>ТРАНСКРИПТ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c.-46A&gt;C</td>\n",
       "      <td>ENST00000371975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c.-46A&gt;C</td>\n",
       "      <td>ENST00000442598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c.-569-152A&gt;C</td>\n",
       "      <td>ENST00000463715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c.-7-39A&gt;C</td>\n",
       "      <td>ENST00000469835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000487700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ГЕН ПОЗИЦИЯ ГЕНЕТИЧЕСКИЙ ВАРИАНТ       ТРАНСКРИПТ\n",
       "0  RAD54L    chr1             c.-46A>C  ENST00000371975\n",
       "1  RAD54L    chr1             c.-46A>C  ENST00000442598\n",
       "2  RAD54L    chr1        c.-569-152A>C  ENST00000463715\n",
       "3  RAD54L    chr1           c.-7-39A>C  ENST00000469835\n",
       "4  RAD54L    chr1                       ENST00000487700"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Сбор всех строк в список\n",
    "rows = list(gene_variant_iterator(vep_df))\n",
    "\n",
    "# Создание DataFrame из списка строк\n",
    "result_df = pd.DataFrame(rows)\n",
    "\n",
    "result_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "40a9b738",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Критерии ACMG\n",
    "acmg_criteria = {\n",
    "    'population': ['PM2', 'BA1', 'BS1', 'BS2'],\n",
    "    'odds_ratio': ['PS4'],\n",
    "    'phenotype': ['PP4'],\n",
    "    'segregation': ['PP1', 'BS4'],\n",
    "    'de_novo': ['PS2', 'PM6'],\n",
    "    'cis_trans': ['PM3', 'BP2'],\n",
    "    'functional': ['PS3', 'BS3'],\n",
    "    'truncation': ['PVS1', 'PM4', 'BP3'],\n",
    "    'amino_acid_substitution': ['PS1', 'PM5', 'PM1', 'PP2', 'BP1'],\n",
    "    'in_silico_predictions': ['PP3', 'BP4', 'BP7'],\n",
    "    'literature': ['BP5', 'PP5', 'BP6']\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cd3eecd",
   "metadata": {},
   "source": [
    "Добавим пока что случайные комбинации классификаций по разным признакам"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "fef43010",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ГЕН</th>\n",
       "      <th>ПОЗИЦИЯ</th>\n",
       "      <th>ГЕНЕТИЧЕСКИЙ ВАРИАНТ</th>\n",
       "      <th>ТРАНСКРИПТ</th>\n",
       "      <th>random_ACMG_combination</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c.-46A&gt;C</td>\n",
       "      <td>ENST00000371975</td>\n",
       "      <td>{'population': 'BA1', 'odds_ratio': 'PS4', 'ph...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c.-46A&gt;C</td>\n",
       "      <td>ENST00000442598</td>\n",
       "      <td>{'population': 'BA1', 'odds_ratio': nan, 'phen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c.-569-152A&gt;C</td>\n",
       "      <td>ENST00000463715</td>\n",
       "      <td>{'population': nan, 'odds_ratio': nan, 'phenot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c.-7-39A&gt;C</td>\n",
       "      <td>ENST00000469835</td>\n",
       "      <td>{'population': nan, 'odds_ratio': nan, 'phenot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000487700</td>\n",
       "      <td>{'population': nan, 'odds_ratio': 'PS4', 'phen...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ГЕН ПОЗИЦИЯ ГЕНЕТИЧЕСКИЙ ВАРИАНТ       ТРАНСКРИПТ  \\\n",
       "0  RAD54L    chr1             c.-46A>C  ENST00000371975   \n",
       "1  RAD54L    chr1             c.-46A>C  ENST00000442598   \n",
       "2  RAD54L    chr1        c.-569-152A>C  ENST00000463715   \n",
       "3  RAD54L    chr1           c.-7-39A>C  ENST00000469835   \n",
       "4  RAD54L    chr1                       ENST00000487700   \n",
       "\n",
       "                             random_ACMG_combination  \n",
       "0  {'population': 'BA1', 'odds_ratio': 'PS4', 'ph...  \n",
       "1  {'population': 'BA1', 'odds_ratio': nan, 'phen...  \n",
       "2  {'population': nan, 'odds_ratio': nan, 'phenot...  \n",
       "3  {'population': nan, 'odds_ratio': nan, 'phenot...  \n",
       "4  {'population': nan, 'odds_ratio': 'PS4', 'phen...  "
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import random\n",
    "import numpy as np\n",
    "# Функция для создания случайной комбинации из списка\n",
    "def generate_random_combination(criteria_list):\n",
    "    if random.random() < 0.7:\n",
    "        return np.nan\n",
    "    return random.choice(criteria_list)\n",
    "\n",
    "# Создание нового столбца с случайными комбинациями\n",
    "result_df['random_ACMG_combination'] = result_df.apply(lambda row: {\n",
    "    category: generate_random_combination(acmg_criteria[category]) for category in acmg_criteria\n",
    "}, axis=1)\n",
    "\n",
    "# Вывод первых нескольких строк датафрейма\n",
    "result_df.head()\n",
    "#result_df.iloc[0]['random_ACMG_combination']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f845eb5",
   "metadata": {},
   "source": [
    "Дальше идет GPT code \n",
    "Надо всё сделать вручную"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "15326163",
   "metadata": {},
   "outputs": [],
   "source": [
    "def has_criterion(combination, criterion):\n",
    "    return criterion in combination.values()\n",
    "\n",
    "def count_strong_criteria(combination):\n",
    "    return sum(criterion.startswith('PS') for criterion in combination.values())\n",
    "\n",
    "def count_moderate_criteria(combination):\n",
    "    return sum(criterion.startswith('PM') for criterion in combination.values())\n",
    "\n",
    "def count_supporting_criteria(combination):\n",
    "    return sum(criterion.startswith('PP') for criterion in combination.values())\n",
    "\n",
    "def count_very_strong_criteria(combination):\n",
    "    return 'PVS1' in combination.values()\n",
    "\n",
    "def count_likely_pathogenic_criteria(combination):\n",
    "    return 'PM1' in combination.values()\n",
    "\n",
    "def count_benign_criteria(combination):\n",
    "    return any(criterion.startswith('BS') for criterion in combination.values())\n",
    "\n",
    "def count_likely_benign_criteria(combination):\n",
    "    return any(criterion.startswith('BP') for criterion in combination.values())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "7ae40255",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def classify_acmg(combination):\n",
    "    # Проверка на патогенность\n",
    "    if isinstance(combination, dict):\n",
    "        if ('PVS1' in combination.values() and\n",
    "            (sum(criterion.startswith('PS') for criterion in combination.values() if isinstance(criterion, str)) >= 1 or\n",
    "             sum(criterion.startswith('PM') for criterion in combination.values() if isinstance(criterion, str)) >= 2 or\n",
    "             (sum(criterion.startswith('PM') for criterion in combination.values() if isinstance(criterion, str)) >= 1 and\n",
    "              sum(criterion.startswith('PP') for criterion in combination.values() if isinstance(criterion, str)) >= 1) or\n",
    "             sum(criterion.startswith('PP') for criterion in combination.values() if isinstance(criterion, str)) >= 2)):\n",
    "            return 'Pathogenic'\n",
    "\n",
    "        if (sum(criterion.startswith('PS') for criterion in combination.values() if isinstance(criterion, str)) >= 2 or\n",
    "            ('PS' in combination.values() and\n",
    "             (sum(criterion.startswith('PM') for criterion in combination.values() if isinstance(criterion, str)) >= 3 or\n",
    "              (sum(criterion.startswith('PM') for criterion in combination.values() if isinstance(criterion, str)) >= 2 and\n",
    "               sum(criterion.startswith('PP') for criterion in combination.values() if isinstance(criterion, str)) >= 2) or\n",
    "              (sum(criterion.startswith('PM') for criterion in combination.values() if isinstance(criterion, str)) >= 1 and\n",
    "               sum(criterion.startswith('PP') for criterion in combination.values() if isinstance(criterion, str)) >= 4)))):\n",
    "            return 'Likely pathogenic'\n",
    "\n",
    "        # Проверка на вероятность патогенности\n",
    "        if (('PVS1' in combination.values() and 'PM1' in combination.values()) or\n",
    "            ('PS' in combination.values() and\n",
    "             (sum(criterion.startswith('PM') for criterion in combination.values() if isinstance(criterion, str)) in [1, 2] and\n",
    "              sum(criterion.startswith('PP') for criterion in combination.values() if isinstance(criterion, str)) >= 2) or\n",
    "             sum(criterion.startswith('PM') for criterion in combination.values() if isinstance(criterion, str)) >= 3 or\n",
    "             (sum(criterion.startswith('PM') for criterion in combination.values() if isinstance(criterion, str)) >= 2 and\n",
    "              sum(criterion.startswith('PP') for criterion in combination.values() if isinstance(criterion, str)) >= 2) or\n",
    "             (sum(criterion.startswith('PM') for criterion in combination.values() if isinstance(criterion, str)) >= 1 and\n",
    "              sum(criterion.startswith('PP') for criterion in combination.values() if isinstance(criterion, str)) >= 4))):\n",
    "            return 'Likely pathogenic'\n",
    "\n",
    "        # Проверка на неопределенное значение\n",
    "        if all(criterion is np.nan for criterion in combination.values()) or \\\n",
    "           (any(criterion.startswith('BS') for criterion in combination.values() if isinstance(criterion, str)) and\n",
    "            any(criterion.startswith('PM') for criterion in combination.values() if isinstance(criterion, str)) or\n",
    "            (any(criterion.startswith('PP') for criterion in combination.values() if isinstance(criterion, str)) and\n",
    "             any(criterion.startswith('PM') for criterion in combination.values() if isinstance(criterion, str)))):\n",
    "            return 'Uncertain significance'\n",
    "\n",
    "        # Проверка на вероятность доброкачественности\n",
    "        if (('PS' in combination.values() and 'BP' in combination.values()) or\n",
    "            (sum(criterion.startswith('BP') for criterion in combination.values() if isinstance(criterion, str)) >= 2)):\n",
    "            return 'Likely benign'\n",
    "\n",
    "        # Проверка на доброкачественное значение\n",
    "        if ('BA1' in combination.values() or\n",
    "            (sum(criterion.startswith('BS') for criterion in combination.values() if isinstance(criterion, str)) >= 2)):\n",
    "            return 'Benign'\n",
    "\n",
    "    return np.nan  # Если не подходит ни одно из условий или тип не является dict\n",
    "\n",
    "# Применение функции к датафрейму\n",
    "result_df['ACMG_classification'] = result_df['random_ACMG_combination'].apply(classify_acmg)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "a99f4eb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ГЕН</th>\n",
       "      <th>ПОЗИЦИЯ</th>\n",
       "      <th>ГЕНЕТИЧЕСКИЙ ВАРИАНТ</th>\n",
       "      <th>ТРАНСКРИПТ</th>\n",
       "      <th>random_ACMG_combination</th>\n",
       "      <th>ACMG_classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c.-46A&gt;C</td>\n",
       "      <td>ENST00000371975</td>\n",
       "      <td>{'population': 'BA1', 'odds_ratio': 'PS4', 'ph...</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c.-46A&gt;C</td>\n",
       "      <td>ENST00000442598</td>\n",
       "      <td>{'population': 'BA1', 'odds_ratio': nan, 'phen...</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c.-569-152A&gt;C</td>\n",
       "      <td>ENST00000463715</td>\n",
       "      <td>{'population': nan, 'odds_ratio': nan, 'phenot...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>c.-7-39A&gt;C</td>\n",
       "      <td>ENST00000469835</td>\n",
       "      <td>{'population': nan, 'odds_ratio': nan, 'phenot...</td>\n",
       "      <td>Uncertain significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RAD54L</td>\n",
       "      <td>chr1</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000487700</td>\n",
       "      <td>{'population': nan, 'odds_ratio': 'PS4', 'phen...</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ГЕН ПОЗИЦИЯ ГЕНЕТИЧЕСКИЙ ВАРИАНТ       ТРАНСКРИПТ  \\\n",
       "0  RAD54L    chr1             c.-46A>C  ENST00000371975   \n",
       "1  RAD54L    chr1             c.-46A>C  ENST00000442598   \n",
       "2  RAD54L    chr1        c.-569-152A>C  ENST00000463715   \n",
       "3  RAD54L    chr1           c.-7-39A>C  ENST00000469835   \n",
       "4  RAD54L    chr1                       ENST00000487700   \n",
       "\n",
       "                             random_ACMG_combination     ACMG_classification  \n",
       "0  {'population': 'BA1', 'odds_ratio': 'PS4', 'ph...              Pathogenic  \n",
       "1  {'population': 'BA1', 'odds_ratio': nan, 'phen...  Uncertain significance  \n",
       "2  {'population': nan, 'odds_ratio': nan, 'phenot...                     NaN  \n",
       "3  {'population': nan, 'odds_ratio': nan, 'phenot...  Uncertain significance  \n",
       "4  {'population': nan, 'odds_ratio': 'PS4', 'phen...       Likely pathogenic  "
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
